File: dotnet/Gitologist/Fetch.cs
using System.Text;
using System.Text.RegularExpressions;
using Gitologist.Types;

namespace Gitologist;

public class FetchResult
{
    public string Remote { get; set; } = null!;
    public List<RefInfo> Refs { get; set; } = new();
}

public class RefInfo
{
    public string Name { get; set; } = null!;
    public string Sha { get; set; } = null!;
}

public class DiscoveredRef
{
    public string Sha { get; set; } = null!;
    public string Ref { get; set; } = null!;
}

public static class Fetch
{
    public static async Task<FetchResult> FetchOrigin(string path, string? remote = null, RemoteOptions? options = null)
    {
        var gitDir = Path.Combine(path, ".git");

        if (!Directory.Exists(gitDir))
        {
            throw new InvalidOperationException("Not a git repository");
        }

        var remoteName = remote ?? "origin";
        var remoteUrl = await Remote.GetRemoteUrl(gitDir, remoteName);

        if (remoteUrl == null)
        {
            return new FetchResult { Remote = remoteName, Refs = new List<RefInfo>() };
        }

        var refs = await DiscoverRefs(remoteUrl, options);
        var result = new FetchResult { Remote = remoteName, Refs = new List<RefInfo>() };

        var wants = new List<string>();
        var haves = new List<string>();

        foreach (var discoveredRef in refs)
        {
            if (discoveredRef.Ref.StartsWith("refs/heads/"))
            {
                var branchName = discoveredRef.Ref.Substring("refs/heads/".Length);
                var localRefPath = Path.Combine(gitDir, "refs", "heads", branchName);

                if (File.Exists(localRefPath))
                {
                    var localSha = (await File.ReadAllTextAsync(localRefPath)).Trim();
                    haves.Add(localSha);
                }

                wants.Add(discoveredRef.Sha);
                result.Refs.Add(new RefInfo { Name = branchName, Sha = discoveredRef.Sha });
            }
        }

        if (wants.Count > 0)
        {
            var objects = await FetchPackfile(remoteUrl, wants, haves, options);
            await StoreObjects(gitDir, objects);
        }

        foreach (var refInfo in result.Refs)
        {
            var remoteRefPath = Path.Combine(gitDir, "refs", "remotes", remoteName, refInfo.Name);
            Directory.CreateDirectory(Path.GetDirectoryName(remoteRefPath)!);
            await File.WriteAllTextAsync(remoteRefPath, refInfo.Sha + "\n");
        }

        return result;
    }

    private static async Task StoreObjects(string gitDir, List<PackObject> objects)
    {
        foreach (var obj in objects)
        {
            await Utils.HashObject(gitDir, obj.Content, obj.Type);
        }
    }

    private static async Task<List<DiscoveredRef>> DiscoverRefs(string remoteUrl, RemoteOptions? options = null)
    {
        var url = new Uri($"{remoteUrl.TrimEnd('/')}/git-upload-pack");

        var requestBody = BuildLsRefsRequest();

        using var client = new HttpClient();
        using var content = new ByteArrayContent(requestBody);
        content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/x-git-upload-pack-request");
        client.DefaultRequestHeaders.Add("Accept", "application/x-git-upload-pack-result");
        client.DefaultRequestHeaders.Add("Git-Protocol", "version=2");

        if (options?.Credentials != null)
        {
            var authString = $"{options.Credentials.Username}:{options.Credentials.Token}";
            var authBytes = Encoding.UTF8.GetBytes(authString);
            var base64Auth = Convert.ToBase64String(authBytes);
            client.DefaultRequestHeaders.Add("Authorization", $"Basic {base64Auth}");
        }

        var response = await client.PostAsync(url, content);
        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException($"Failed to discover refs: {(int)response.StatusCode} {response.ReasonPhrase}");
        }

        var data = await response.Content.ReadAsByteArrayAsync();
        var lines = Packfile.DecodePktLines(data);

        var refs = new List<DiscoveredRef>();

        foreach (var line in lines)
        {
            var trimmed = line.Trim();
            if (string.IsNullOrEmpty(trimmed)) continue;

            var parts = trimmed.Split(new[] { ' ' }, 2);
            if (parts.Length >= 2)
            {
                var sha = parts[0];
                if (sha.Length == 40 && sha.All(c => char.IsLetterOrDigit(c)))
                {
                    var refName = parts[1].Trim().Split('\0')[0];
                    refs.Add(new DiscoveredRef { Sha = sha, Ref = refName });
                }
            }
        }

        return refs;
    }

    private static byte[] BuildLsRefsRequest()
    {
        var lines = new List<byte[]>();

        lines.Add(Packfile.EncodePktLine("command=ls-refs\n"));
        lines.Add(Encoding.UTF8.GetBytes("0001"));
        lines.Add(Packfile.EncodePktLine("symrefs\n"));
        lines.Add(Packfile.EncodePktLine("peel\n"));
        lines.Add(Packfile.EncodePktLine("ref-prefix refs/heads/\n"));
        lines.Add(Packfile.EncodePktLine(null));

        return lines.SelectMany(x => x).ToArray();
    }

    private static byte[] ExtractPackfileFromSideband(byte[] data)
    {
        var offset = 0;
        var packfileData = new List<byte>();

        while (offset < data.Length)
        {
            if (offset + 4 > data.Length)
            {
                break;
            }

            var hexLength = Encoding.ASCII.GetString(data, offset, 4);

            if (hexLength == "0000")
            {
                offset += 4;
                continue;
            }

            if (hexLength == "0001")
            {
                offset += 4;
                continue;
            }

            if (!int.TryParse(hexLength, System.Globalization.NumberStyles.HexNumber, null, out var length))
            {
                break;
            }

            if (length <= 0 || offset + length > data.Length)
            {
                break;
            }

            var payload = new byte[length - 4];
            Array.Copy(data, offset + 4, payload, 0, length - 4);

            if (payload.Length > 0)
            {
                var channel = payload[0];
                if (channel == 1)
                {
                    packfileData.AddRange(payload.Skip(1));
                }
                else if (channel == 3)
                {
                    var errorMsg = Encoding.UTF8.GetString(payload, 1, payload.Length - 1);
                    Console.Error.WriteLine($"Git error: {errorMsg}");
                }
            }

            offset += length;
        }

        return packfileData.ToArray();
    }

    private static async Task<List<PackObject>> FetchPackfile(string remoteUrl, List<string> wants, List<string> haves, RemoteOptions? options = null)
    {
        var url = new Uri($"{remoteUrl.TrimEnd('/')}/git-upload-pack");

        var requestBody = BuildFetchRequest(wants, haves);

        using var client = new HttpClient();
        using var content = new ByteArrayContent(requestBody);
        content.Headers.ContentType = new System.Net.Http.Headers.MediaTypeHeaderValue("application/x-git-upload-pack-request");
        client.DefaultRequestHeaders.Add("Accept", "application/x-git-upload-pack-result");
        client.DefaultRequestHeaders.Add("Git-Protocol", "version=2");

        if (options?.Credentials != null)
        {
            var authString = $"{options.Credentials.Username}:{options.Credentials.Token}";
            var authBytes = Encoding.UTF8.GetBytes(authString);
            var base64Auth = Convert.ToBase64String(authBytes);
            client.DefaultRequestHeaders.Add("Authorization", $"Basic {base64Auth}");
        }

        var response = await client.PostAsync(url, content);
        if (!response.IsSuccessStatusCode)
        {
            throw new InvalidOperationException($"Failed to fetch packfile: {(int)response.StatusCode} {response.ReasonPhrase}");
        }

        var data = await response.Content.ReadAsByteArrayAsync();
        var packfileData = ExtractPackfileFromSideband(data);

        if (packfileData.Length == 0)
        {
            return new List<PackObject>();
        }

        return Packfile.ParsePackfile(packfileData);
    }

    private static byte[] BuildFetchRequest(List<string> wants, List<string> haves)
    {
        var lines = new List<byte[]>();

        lines.Add(Packfile.EncodePktLine("command=fetch\n"));
        lines.Add(Encoding.UTF8.GetBytes("0001"));

        foreach (var want in wants)
        {
            lines.Add(Packfile.EncodePktLine($"want {want}\n"));
        }

        foreach (var have in haves)
        {
            lines.Add(Packfile.EncodePktLine($"have {have}\n"));
        }

        lines.Add(Packfile.EncodePktLine("done\n"));
        lines.Add(Packfile.EncodePktLine(null));

        return lines.SelectMany(x => x).ToArray();
    }
}

File: web/src/fetch.ts
import { existsSync } from "node:fs";
import { readFile, writeFile, mkdir } from "node:fs/promises";
import { join, dirname } from "node:path";

import { encodePktLine, decodePktLines, parsePackfile, type PackObject } from "./packfile.ts";
import type { FetchResult } from "./types/FetchResult.ts";
import { hashObjectBuffer } from "./utils.ts";

export async function fetchFromRemote(path: string, remote?: string): Promise<FetchResult> {
	const gitDir = join(path, ".git");

	if (!existsSync(gitDir)) {
		throw new Error("Not a git repository");
	}

	const remoteName = remote || "origin";
	const remoteUrl = await getRemoteUrl(gitDir, remoteName);

	if (!remoteUrl) {
		return {
			remote: remoteName,
			refs: [],
		};
	}

	const refs = await discoverRefs(remoteUrl);
	const result: FetchResult = {
		remote: remoteName,
		refs: [],
	};

	const wants: string[] = [];
	const haves: string[] = [];

	for (const ref of refs) {
		if (ref.ref.startsWith("refs/heads/")) {
			const branchName = ref.ref.slice("refs/heads/".length);
			const localRefPath = join(gitDir, "refs", "heads", branchName);

			if (existsSync(localRefPath)) {
				const localSha = (await readFile(localRefPath, "utf-8")).trim();
				haves.push(localSha);
			}

			wants.push(ref.sha);
			result.refs.push({
				name: branchName,
				sha: ref.sha,
			});
		}
	}

	if (wants.length > 0) {
		const objects = await fetchPackfile(remoteUrl, wants, haves);
		await storeObjects(gitDir, objects);
	}

	for (const ref of result.refs) {
		const remoteRefPath = join(gitDir, "refs", "remotes", remoteName, ref.name);
		await mkdir(dirname(remoteRefPath), { recursive: true });
		await writeFile(remoteRefPath, ref.sha + "\n", "utf-8");
	}

	return result;
}

async function storeObjects(gitDir: string, objects: PackObject[]): Promise<void> {
	for (const obj of objects) {
		const type = obj.type;
		await hashObjectBuffer(gitDir, obj.content, type);
	}
}

async function getRemoteUrl(gitDir: string, remoteName: string): Promise<string | null> {
	const configPath = join(gitDir, "config");

	if (!existsSync(configPath)) {
		return null;
	}

	const configContent = await readFile(configPath, "utf-8");
	const lines = configContent.split("\n");

	let inRemoteSection = false;
	let currentRemote = "";

	for (const line of lines) {
		const trimmed = line.trim();

		const sectionMatch = trimmed.match(/^\[remote\s+"([^"]+)"\]$/);
		if (sectionMatch) {
			inRemoteSection = true;
			currentRemote = sectionMatch[1];
			continue;
		}

		if (inRemoteSection && currentRemote === remoteName) {
			const urlMatch = trimmed.match(/^url\s*=\s*(.+)$/);
			if (urlMatch) {
				return urlMatch[1].trim();
			}
		}

		if (trimmed.startsWith("[") && !trimmed.startsWith("[remote")) {
			inRemoteSection = false;
		}
	}

	return null;
}

interface DiscoveredRef {
	sha: string;
	ref: string;
}

async function discoverRefs(remoteUrl: string): Promise<DiscoveredRef[]> {
	const url = new URL("info/refs", remoteUrl);
	url.searchParams.set("service", "git-upload-pack");

	const response = await fetch(url.toString(), {
		method: "GET",
		headers: {
			Accept: "application/x-git-upload-pack-advertisement",
			"Git-Protocol": "version=2",
		},
	});

	if (!response.ok) {
		throw new Error(`Failed to discover refs: ${response.status} ${response.statusText}`);
	}

	const buffer = Buffer.from(await response.arrayBuffer());
	const lines = decodePktLines(buffer);

	const refs: DiscoveredRef[] = [];
	let started = false;

	for (const line of lines) {
		if (line.includes("# service=git-upload-pack")) {
			started = true;
			continue;
		}

		if (!started) continue;
		if (line === "") continue;

		const parts = line.split(" ");
		if (parts.length >= 2 && parts[0].match(/^[0-9a-f]{40}$/)) {
			refs.push({
				sha: parts[0],
				ref: parts[1].split("\0")[0],
			});
		}
	}

	if (refs.length > 0) {
		return refs;
	}

	return await discoverRefsV2(remoteUrl);
}

async function discoverRefsV2(remoteUrl: string): Promise<DiscoveredRef[]> {
	const uploadUrl = new URL("git-upload-pack", remoteUrl);

	const requestBody = buildLsRefsRequest();

	const response = await fetch(uploadUrl.toString(), {
		method: "POST",
		headers: {
			"Content-Type": "application/x-git-upload-pack-request",
			Accept: "application/x-git-upload-pack-result",
			"Git-Protocol": "version=2",
		},
		body: requestBody,
	});

	if (!response.ok) {
		throw new Error(`Failed to discover refs v2: ${response.status} ${response.statusText}`);
	}

	const buffer = Buffer.from(await response.arrayBuffer());
	const lines = decodePktLines(buffer);

	const refs: DiscoveredRef[] = [];

	for (const line of lines) {
		const trimmed = line.trim();
		if (trimmed === "") continue;

		const parts = trimmed.split(" ");
		if (parts.length >= 2 && parts[0].match(/^[0-9a-f]{40}$/)) {
			refs.push({
				sha: parts[0],
				ref: parts[1].trim().split("\0")[0],
			});
		}
	}

	return refs;
}

function buildLsRefsRequest(): Buffer {
	const lines: Buffer[] = [];

	lines.push(encodePktLine("command=ls-refs\n"));
	lines.push(encodePktLine("0001"));
	lines.push(encodePktLine("symrefs\n"));
	lines.push(encodePktLine("peel\n"));
	lines.push(encodePktLine("ref-prefix refs/heads/\n"));
	lines.push(encodePktLine(null));

	return Buffer.concat(lines);
}

function extractPackfileFromSideband(data: Buffer): Buffer {
	let offset = 0;
	const packfileData: Buffer[] = [];

	while (offset < data.length) {
		if (offset + 4 > data.length) {
			break;
		}

		const hexLen = data.slice(offset, offset + 4).toString("ascii");

		if (hexLen === "0000") {
			offset += 4;
			continue;
		}

		if (hexLen === "0001") {
			offset += 4;
			continue;
		}

		const length = parseInt(hexLen, 16);
		if (length <= 0 || offset + length > data.length) {
			break;
		}

		const payload = data.slice(offset + 4, offset + length);

		if (payload.length > 0) {
			const channel = payload[0];
			if (channel === 1) {
				packfileData.push(payload.slice(1));
			} else if (channel === 3) {
				const errorMsg = payload.slice(1).toString("utf-8");
				console.error(`Git error: ${errorMsg}`);
			}
		}

		offset += length;
	}

	return Buffer.concat(packfileData);
}

async function fetchPackfile(
	remoteUrl: string,
	wants: string[],
	haves: string[],
): Promise<PackObject[]> {
	const url = new URL("git-upload-pack", remoteUrl);

	const requestBody = buildFetchRequestV2(wants, haves);

	const response = await fetch(url.toString(), {
		method: "POST",
		headers: {
			"Content-Type": "application/x-git-upload-pack-request",
			Accept: "application/x-git-upload-pack-result",
			"Git-Protocol": "version=2",
		},
		body: requestBody,
	});

	if (!response.ok) {
		throw new Error(`Failed to fetch packfile: ${response.status} ${response.statusText}`);
	}

	const buffer = Buffer.from(await response.arrayBuffer());

	const packfileData = extractPackfileFromSideband(buffer);
	if (packfileData.length === 0) {
		return [];
	}

	return parsePackfile(packfileData);
}

function buildFetchRequestV2(wants: string[], haves: string[]): Buffer {
	const lines: Buffer[] = [];

	lines.push(encodePktLine("command=fetch\n"));
	lines.push(encodePktLine("0001"));

	for (const want of wants) {
		lines.push(encodePktLine(`want ${want}\n`));
	}

	for (const have of haves) {
		lines.push(encodePktLine(`have ${have}\n`));
	}

	lines.push(encodePktLine("done\n"));
	lines.push(encodePktLine(null));

	return Buffer.concat(lines);
}
